Implementation of Rest-Support

ManuelMuehlberger opened the pull request:
Both IssueReviewsCacheLoader and AdminServlet can now establish a Rest-connection to Gerrit with the help of a modified gerrit-events dependency.

ManuelMuehlberger committed:
first iteration of Rest-support

Both IssueReviewsCacheLoader and AdminServlet can establish a Rest-connection with the help of a modified gerrit-events dependency

f10mlupfer commented on pom.xml:
```diff
@@ -137,30 +166,21 @@
+            <groupId>com.sonymobile.tools.gerrit</groupId>
             <artifactId>gerrit-events</artifactId>
-            <version>1.99.1</version>
+            <version>2.20.1</version>
```
für das lokale Testen muss man hier -SNAPSHOT angeben, da die 2.20.1 noch nicht released ist. Wahrscheinlich ist das erstmal besser so, bis wir das final released haben

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/workflow/function/ApprovalFunction.java:
```diff
@@ -25,7 +25,7 @@
 import com.opensymphony.module.propertyset.PropertySet;
 import com.opensymphony.workflow.WorkflowException;
-import com.sonyericsson.hudson.plugins.gerrit.gerritevents.GerritQueryException;
+import com.sonymobile.tools.gerrit.gerritevents.GerritQueryException;
```
hat nichts mit der Zeile zu tun, aber beim bauen bekomme ich beim javadoch einen Fehler, dass die <tt> Tags nicht supported sind => <tt>gerrit review [ChangeId] ...</tt> (Zeile 40)
?

ManuelMuehlberger replied:
Kann ich jetzt leider nicht reproduzieren... Kannst du mir da noch etwas mehr Detail geben?

f10mlupfer replied:
ich habe einfach mvn clean install auf dem Projekt aufgerufen. Hat es evtl. etwas mit der maven Version zu tun?

f10mlupfer replied:
ich habe einfach mvn clean install auf dem Projekt aufgerufen. Hat es evtl. etwas mit der maven Version zu tun?

ManuelMuehlberger replied:
Ich habe jetzt mal ne Weile mit verschiedenen Maven und Atlassian-maven Versionen getestet, ich komme aber nur zu einem von zwei Ergebnissen: 1. mit Atlassian-mvn funktioniert alles (Spotbugs hat sich noch beschwert, dass die mvn version nichtmehr unterstützt wird, ich kann mir aber nicht vorstellen, dass das Einfluss auf javadoc hat -> version etwas downgegradet) oder 2. mit normalen Maven kann er die besonderen atlassian-Dependencies nicht finden

f10mlupfer replied:
also wenn es so durchläuft ist es ja erstmal für mich ok!

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/GerritCommand.java:
```diff
@@ -69,6 +67,7 @@ private String getCommand(GerritChange change, String args) {
         return String.format("%s %s,%s %s", BASE_COMMAND, change.getNumber(), change.getPatchSet().getNumber(), args);
     }
 
+    //Rest support not needed, unused method
```
wenn die Methode nicht verwendet wird, warum wird sie dann nicht entfernt (und ggf. weitere, die dadurch evtl. unnötig werden)?

ManuelMuehlberger replied:
Keine Ahnung, warum die als depreciated markiert wurde. Die Methode wird benutzt, es gibt außerdem auch keine andere Methode, welche diese ablösen würde. Markierung entfernt

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -192,23 +194,63 @@ private Map<String, Object> handlePost(HttpServletRequest req, HttpServletRespon
-            map.put("testError", e.getMessage());
+            try {
+                //trying to fetch the newest change
+                List<JSONObject> s = query.queryJava("limit:1", false, false, false);
```
mit "s" wird nichts gemacht

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -192,23 +194,63 @@ private Map<String, Object> handlePost(HttpServletRequest req, HttpServletRespon
+                            return configuration.getHttpUsername();
+                        }
+                    };
+                    //() -> ""; lambdas are tripping up CheckStyle
```
kann das raus oder was ist damit?

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -192,23 +194,63 @@ private Map<String, Object> handlePost(HttpServletRequest req, HttpServletRespon
-            map.put("testError", "not configured");
-            return;
+        if(configuration.getPreferRestConnection()) {
+            //TODO: better URL-check
+            //https://toolsqa.com/rest-assured/rest-api-test-using-rest-assured/
```
hat sich das TODO erledigt?

ManuelMuehlberger replied:
Habe ich mal umgelabelt zu "future idea". Besseres Url checking wäre ein Nice-to-have, das bisherige, eher rudimentäre, sollte aber auch OK sein.

f10mlupfer replied:
ok. Ansonsten könntest du es so drin lassen und ggf. im Wiki dazu noch einen zusätzlichen Absatz bzgl. Verbesserungen/Improvements/TODOs

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -37,25 +36,24 @@
 public class AdminServlet extends HttpServlet {
     private static final long serialVersionUID = -9175363090552720328L;
-    private static final Logger log = LoggerFactory.getLogger(AdminServlet.class);
+    protected static Logger log = LoggerFactory.getLogger(AdminServlet.class);
```
muss der protected sein? die Klasse wird doch nicht abgeleitet

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -37,25 +36,24 @@
-import javax.servlet.http.HttpServlet;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
+//class is described in detail here:    https://developer.atlassian.com/server/framework/atlassian-sdk/creating-an-admin-configuration-form/
```
hier würde einen richtigen JavaDoc Kommentar bevorzugen, bspw.
/**
  \* @see https://developer.atlassian.com/server/framework/atlassian-sdk/creating-an-admin-configuration-form/
 */

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/GerritConfigurationImpl.java:
```diff
@@ -62,11 +65,24 @@ public String getHttpUsername() {
+
+    {
+        return Boolean.parseBoolean((String) settings.get(FIELD_PREFER_REST_CONNECTION));
+    }
```
Formatierung

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/GerritConfigurationImpl.java:
```diff
@@ -151,6 +167,15 @@ && getSshPrivateKey() != null
+                && !Strings.isNullOrEmpty(getHttpPassword());
+    }
+
+
```
Unnötige Leerzeilen

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/IssueReviewsCacheLoader.java:
```diff
@@ -45,22 +47,59 @@ public List<GerritChange> load(@Nonnull String key) {
+        if(configuration.getPreferRestConnection())
+        {
+            if (!configuration.isHttpValid()) {
+                // return Collections.emptyList();
```
kann das raus?

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/IssueReviewsCacheLoader.java:
```diff
@@ -45,22 +47,59 @@ public List<GerritChange> load(@Nonnull String key) {
+                            return configuration.getHttpUsername();
+                        }
+                    };
+                    //() -> ""; lambdas are tripping up CheckStyle
```
und das?

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/IssueReviewsCacheLoader.java:
```diff
@@ -45,22 +47,59 @@ public List<GerritChange> load(@Nonnull String key) {
+        else
+        {
+            if (!configuration.isSshValid()) {
+                // return Collections.emptyList();
```
?

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/dto/GerritApproval.java:
```diff
@@ -14,16 +14,17 @@
 import com.atlassian.jira.user.ApplicationUser;
-import com.sonyericsson.hudson.plugins.gerrit.gerritevents.dto.attr.Approval;
+import com.sonymobile.tools.gerrit.gerritevents.dto.attr.Account;
+import com.sonymobile.tools.gerrit.gerritevents.dto.attr.Approval;
```
wird nirgends verwendet

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/dto/GerritApproval.java:
```diff
@@ -97,7 +98,9 @@ public static String getUpgradedLabelType(String type) {
      */
-    public String getBy() {
+    public String getByString() {
+        //return by;
```
?

ManuelMuehlberger replied:
GerritApproval erbt aus einer Klasse von gerrit-events. Dort wurde der Rückgabewert bei einer der neueren Versionen von String auf eine "Account"klasse geändert, was die originale Methode zerschossen hat... Da wir aber manchmal nur den Namen des Approvers brauchen, habe ich einfach die Signatur geändert. (jetzt auf getByName() damit es klarer ist)

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/dto/GerritChange.java:
```diff
@@ -85,6 +106,120 @@ public void fromJson(JSONObject json) {
+            /*if (!scheme.contains(com.sonymobile.tools.gerrit.gerritevents.dto.GerritEventKeys.URL_PREFIX)) {
+                scheme = com.sonymobile.tools.gerrit.gerritevents.dto.GerritEventKeys.URL_PREFIX + scheme;
+            }
+             */
```
ist das etwas, was wir für die Zukunft im Auge behalten sollten oder kann das raus?

ManuelMuehlberger replied:
Code von dem Fork, der eh nicht funktioniert hat und ich durch meinen URL-Parser ersetzt habe. Kann raus.

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/dto/GerritChange.java:
```diff
@@ -93,7 +228,7 @@ public GerritPatchSet getPatchSet() {
     }
 
-    public String getStatus() {
+    public String getStatusString() {
```
wird die Methode überhaupt verwendet. Wenn nicht raus, ansonsten, was spricht gegen getStatus()?

ManuelMuehlberger replied:
wie oben, veränderte Superklasse. Kann aber raus.

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/tabpanel/GerritReviewsTabPanel.java:
```diff
@@ -29,7 +29,8 @@
 import com.atlassian.sal.api.ApplicationProperties;
 import com.atlassian.sal.api.message.I18nResolver;
-import com.sonyericsson.hudson.plugins.gerrit.gerritevents.GerritQueryException;
+//import com.atlassian.sal.api.user.UserManager;
```
kann raus

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/webpanel/GerritReviewsIssueAgilePanel.java:
```diff
@@ -20,36 +20,41 @@
 
 @SuppressWarnings("unchecked")
-public class GerritReviewsIssueAgilePanel extends AbstractJiraContextProvider {
+public class GerritReviewsIssueAgilePanel extends AbstractJiraContextProvider
+{
```
Formatierung sollte einheitlich sein, ggf. über alles mal drüber und neu formatieren

ManuelMuehlberger replied:
Hab alles mal durch den Formatter laufen lassen

f10mlupfer commented on src/main/resources/gerrit-events/pom.xml:
```diff
@@ -0,0 +1,340 @@
+    </developers>
+
+    <dependencies>
+        <!-- <dependency>
+            <groupId>org.apache.mina</groupId>
+            <artifactId>mina-core</artifactId>
+            <version>1.1.7</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.sshd</groupId>
+            <artifactId>apache-sshd</artifactId>
+            <version>0.3.0</version>
+        </dependency>-->
```
können dieses auskommentierten Dependencies raus?

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/adminui/AdminServletTest.java:
```diff
@@ -0,0 +1,172 @@
+     * Initialise the mock-REST-API, call
+     * {@link com.sonymobile.tools.gerrit.gerritevents.GerritQueryHandlerHttp#queryJava(String, boolean, boolean, boolean)}.
+     */
+    @Before
+    public void setupMockServer() {
+        wireMockServer = new WireMockServer(new WireMockConfiguration().port(port));
+        //.usingFilesUnderDirectory("src/test/"));
+        wireMockServer.start();
+    }
```
kann man das in ein @Before mit dem oberen packen? Hier macht es ggf. nichts aus, aber in manchen Fällen ist es sinnvoll zu wissen, dass erst das eine, dann das andere ausgeführt wird. Bei zwei Methoden weiß man das nie so genau, welche als erste ausgeführt wird...

ManuelMuehlberger replied:
habe jetzt bei einem die Annotation entfernt, wird am Ende der anderen aufgerufen.

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/adminui/AdminServletTest.java:
```diff
@@ -0,0 +1,172 @@
+    @Before
+    public void setupMockServer() {
+        wireMockServer = new WireMockServer(new WireMockConfiguration().port(port));
+        //.usingFilesUnderDirectory("src/test/"));
```
das braucht es dann wahrscheinlich auch nicht?

ManuelMuehlberger replied:
Meinst du damit den Kommentar?

f10mlupfer replied:
das hier //.usingFilesUnderDirectory("src/test/"));

ManuelMuehlberger committed:
add url linking back to gerrit

the fromJsonRest() function has been modified and will now create links referring back to gerrit's view of a change correctly.

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/adminui/AdminServletTest.java:
```diff
@@ -0,0 +1,172 @@
+        adminServlet.performConnectionTest(gerritConfiguration, map);
+        assertFalse((Boolean) map.get("testResult"));
+        assertTrue(((String) map.get("testError")).endsWith("(401)"));
+        System.out.println(map.toString());
```
die System.out Befehle würde ich rausnehmen (zumindest auskommentieren, in den Tests als auskommentiert fände ich auch ok.)

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/adminui/AdminServletTest.java:
```diff
@@ -0,0 +1,172 @@
+        gerritConfiguration.setHttpPassword(password);
+
+        adminServlet.performConnectionTest(gerritConfiguration, map);
+        System.out.println("Testresult: " + map.get("testResult"));
```
hier würde ich sagen, System.out ist ok, da man das wahrscheinlich nur lokal einkommentiert und den Test kurz gegen einen richtigen Server laufen lässt.

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/data/IssueReviewsCacheLoaderTest.java:
```diff
@@ -0,0 +1,424 @@
+    public void setupMockServer() {
+        wireMockServer = new WireMockServer(new WireMockConfiguration().port(port)
+                .usingFilesUnderDirectory("src/test/resources/com/sonymobile/tools/gerrit/gerritevents/SampleResponses"));
+        //.usingFilesUnderDirectory("src/test/"));
```
kann raus

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/adminui/AdminServletTest.java:
```diff
@@ -0,0 +1,172 @@
+     * Useful for testing against real servers
+     */
+    //@Test
+    public void testArbitraryServer()
+    {
```
könntest du mal schauen, ob du den Code durchgängig gleich formatiert bekommst? ggf. auch in Richtung "wie er im Original" formatiert ist, ansonsten würde es noch mehr Änderungen geben.

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/data/IssueReviewsCacheLoaderTest.java:
```diff
@@ -0,0 +1,424 @@
+                .willReturn(aResponse().withHeader("Content-Type", "text/plain")
+                        .withStatus(200)
+                        .withBody(")]}'\n" +
+                                "[{\"id\":\"test0%2Ftest0.test~master~Id145cfc4s2d6fek18c3539gbi8me1d3e37501568\",\"project\":\"test0/test0.test\",\"branch\":\"master\",\"attention_set\":{\"309\":{\"account\":{\"_account_id\":309},\"last_update\":\"2022-03-24 09:59:36.000000000\",\"reason\":\"Username0 replied on the change\"}},\"hashtags\":[],\"change_id\":\"Id145cfc4s2d6fek18c3539gbi8me1d3e3750 ...
```
kann man das nicht aus der Datei laden?

ManuelMuehlberger replied:
Directory der Dateien musste geändert werden, geht jetzt.

f10mlupfer commented on src/test/java/com/meetme/plugins/jira/gerrit/data/dto/GerritApprovalTest.java:
```diff
@@ -58,7 +58,7 @@ private static void setUpJson(JSONObject obj) {
     }
 
     private static void assertFull(GerritApproval obj) {
```
muss die static sein?

f10mlupfer reviewed (changes requested):
nicht von der Anzahl der Kommentare erschlagen lassen. Aber noch ein paar Dinge...

ManuelMuehlberger committed:
resolved code review

ManuelMuehlberger committed:
remove calculation of average load time

Calculation of average query time is now optional

f10mlupfer reviewed (changes requested):
muss mir die Änderungen noch anschauen. Bis auf die noch offenen schaut es ja jetzt ganz gut aus 👍

ManuelMuehlberger committed:
edited readme.md to reflect our changes

ManuelMuehlberger committed:
CI badge in readme

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/adminui/AdminServlet.java:
```diff
@@ -37,34 +36,35 @@
     private static final String CONTENT_TYPE = "text/html;charset=utf-8";
 
     private static final String FIELD_ACTION = "action";
     private static final String ACTION_SAVE = "save";
```
hab gerade noch gesehen, dass das ACTION_SAVE nicht verwendet wird

f10mlupfer commented on pom.xml:
```diff
@@ -60,20 +50,31 @@
             <artifactId>jira-api</artifactId>
             <version>${jira.version}</version>
             <scope>provided</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-api</artifactId>
+                </exclusion>
+            </exclusions>
```
noch mal kurz zu den ganzen exclusions, im speziellen org.slf4j, da das ja eigentlich immer exluded wird. Ist das wirklich notwendig? Haben wir das gemacht, weil wir ansonsten viele unterschiedlichen Versionen davon haben?

f10mlupfer commented on src/main/java/com/meetme/plugins/jira/gerrit/data/dto/GerritChange.java:
```diff
@@ -93,10 +238,6 @@ public GerritPatchSet getPatchSet() {
         return patchSet;
     }
 
-    public String getStatus() {
-        return status;
-    }
-
```
status wird auch nicht mehr verwendet?

f10mlupfer reviewed (changes requested):
könntest du nochmal über den ganzen Code drüber schauen (ok, zugegeben, der war ja vorher schon nicht schön), und mal prüfen, was es noch an Code gibt, welcher nicht verwerndet wird oder unnötige Annotations bspw. (@SuppressWarnings("unchecked") in GerritReviewsIssueAgilePanel) gibt und aufräumen?
Ich finde, in diesem Zug könnten wir da gerade noch ein wenig aufräumen

ManuelMuehlberger committed:
pom.xml bearbeitet

-finale gerrit-events dependency
-boolean Methode invertiert
-pom.xml aktualisiert
-

ManuelMuehlberger committed:
Code clean-up

-unnötige Warnings entfernt

ManuelMuehlberger committed:
testing new pipeline

ManuelMuehlberger committed:
Update pom.xml

ManuelMuehlberger committed:
Credential was refactored, change package

ManuelMuehlberger committed:
Fix pipeline
